### DISCUSSION TITLE
Selecting a microcontroller for as many Virtual COMports

### DISCUSSION THREAD

**Discussion by FunDeckHermit** [hathach/tinyusb#1606] [Q&A]
I'm looking for a micro-controller and I was hoping that you have some insights. My endgoal is to have as many Virtual Comports as possible. 15 seems to be the theoretical maximum, 16 is what I actually need.

1. 32 USB endpoints
2. Somewhat new
3. A secondary USB host port
4. ARM Cortex M0 or Cortex M4

The ports are only used to output data, are endpoins fixed as IN/OUT or are they adjustable? Is the INPUT endpoint even optional like the interrupt?

**Comment by pigrew** [comment]
I've never been able to find a specification of "Virtual COM ports" (It uses USB CDC ACM subclass, which doesn't seem to specify endpoint types....), but in practice they always have both a bulk IN and bulk OUT.

The USB Spec has 4-bit endpoint addresses (16 bidirectional endpoints), but EP0 needs to be the control endpoint, yielding a max of 15 bulk endpoints.  Even if you only needed OUT endpoints, it wouldn't work since each endpoint address has one IN and one OUT... the protocol doesn't allow to remap IN endpoints for OUT.... (there are four address bits and one direction bit.)

The RP2040 appears to support 16 hardware endpoints in device mode. In addition, there is a USB implementation for the RP2040 PIO peripherals... so you perhaps could use a USB hub with two USB devices running on a single RP2040.

Other MCU such as many STM32 have multiple USB peripherals (generally one high-speed-capable, and one full speed), but they only have 8 bidirectional endpoints each, which isn't enough.
